### ISSUE TITLE
connect_timeout after enough http errors

### ISSUE THREAD

**Issue by IceDragon200** [edgurgel/httpoison#414]
Sorry if the title isn't descriptive enough, but the jist of it is:

When HTTPoison (or hackney) has enough errors the pool gets completely exhausted and all subsequent requests return checkout_timeout

Test:

```elixir
for _ <- 0..30 do 
  IO.inspect HTTPoison.get("http://localhost:3233")
  :ok # the IO.inspect above is to log the errors as they return
end
```

You may need to run the above at least twice before it starts happening

Example of the result:

```elixir
{:error, %HTTPoison.Error{id: nil, reason: :econnrefused}}
{:error, %HTTPoison.Error{id: nil, reason: :econnrefused}}
... # truncated for brevity
{:error, %HTTPoison.Error{id: nil, reason: :checkout_timeout}}
{:error, %HTTPoison.Error{id: nil, reason: :checkout_timeout}}
{:error, %HTTPoison.Error{id: nil, reason: :checkout_timeout}}
{:error, %HTTPoison.Error{id: nil, reason: :checkout_timeout}}
{:error, %HTTPoison.Error{id: nil, reason: :checkout_timeout}}
{:error, %HTTPoison.Error{id: nil, reason: :checkout_timeout}}
{:error, %HTTPoison.Error{id: nil, reason: :checkout_timeout}}
{:error, %HTTPoison.Error{id: nil, reason: :checkout_timeout}}
{:error, %HTTPoison.Error{id: nil, reason: :checkout_timeout}}
{:error, %HTTPoison.Error{id: nil, reason: :checkout_timeout}}
{:error, %HTTPoison.Error{id: nil, reason: :checkout_timeout}}
{:error, %HTTPoison.Error{id: nil, reason: :checkout_timeout}}
```

All subsequent request will have a checkout_timeout

I'm not sure if this an HTTPoison or hackney issue at the moment

This test was done in a completely new project (no additional configuration) with only httpoison installed (and it's dependencies)

Versions:

```
httpoison 1.6.2
hackney 1.16.0
```

**Comment by nateless** [comment]
We get the same error, it looks like processes aren't killed by timeout and pool is exhausted:

`iex(30)> :hackney_pool.get_stats :default
[name: :default, max: 50, in_use_count: 50, free_count: 0, queue_count: 0]`

**Comment by neoamos** [comment]
Looks like it could have been introduced with version 1.16 of hackney.  Several changes were made to the pool in this version:
https://github.com/benoitc/hackney/commit/8bb529edd105c40859e6a5599324b6129ba8993c

**Comment by wrren** [comment]
We're seeing this issue in one of our production apps and have downgraded hackney to version 1.15.2 through an override.

**Comment by atomkirk** [comment]
and now we're all in a pickle because we can't upgrade to OTP 23 without hackney 1.16…

**Comment by overture8** [comment]
Has this issue been resolved? I'm getting it using version 1.7.0 in production. Should I just downgrade? Thanks!

**Comment by nateless** [comment]
@overture8 no, its not. the issue is with hackney (https://github.com/benoitc/hackney/issues/643). I would recommend catching `: checkout_timeout ` and restart the pool.

**Comment by whossname** [comment]
code for natelesses suggestion for anyone else who isn't too familiar with hackney's API

```
for _ <- 0..100 do 
  result = HTTPoison.get("http://localhost:3233")
  
  result
  |> case do
     {:error, %HTTPoison.Error{reason: :checkout_timeout}} ->
        IO.inspect("fixing timeout")
        :hackney_pool.stop_pool(:default)
     _ -> nil
  end

  result
end
```

**Comment by IceDragon200** [comment]
Just an update for those who don't really track hackney's issues

https://github.com/benoitc/hackney/pull/661

That PR should resolve the issue, if you're willing to test it out, let everyone know over there if it works for you, the more eyes we have on the issue, the faster we can get it merged and pushed upstream (I think)

**Comment by damonvjanis** [comment]
It looks like this was included in the 1.17 release, so if you update hackney it will fix the issue.

https://github.com/benoitc/hackney/blob/master/NEWS.md

**Comment by bukenbp8** [comment]
When using pools, one :timeout produces :checkout_timeout on all following requests (more than 200). Pool size is default (as I understand is 50).

If I set pool: false everything works as expected. It would be nice to use pools for performance reasons...

httpoison 1.8
hackney 1.17
